src/main/test/ApplicationUtilsTests.cpp

```diff
@@ -146,3 +150,197 @@ TEST_CASE("offline self-check works", "[applicationutils][selfcheck]")
+            validator->getHistoryManager().getCheckpointFrequency();
+
+        uint32_t selectedLedger = 0;
+        std::string selectedHash;
+
+        // Make sure validator publishes something
+        simulation->crankUntil(
+            [&]() {
+                bool loadDone = loadGenDone.count() > currLoadGenCount;
+                auto lcl =
+                    validator->getLedgerManager().getLastClosedLedgerHeader();
+                // Pick some ledger in the second checkpoint to run catchup
+                // against later
+                if (lcl.header.ledgerSeq < 2 * checkpoint)
+                {
+                    selectedLedger = lcl.header.ledgerSeq;
+                    selectedHash = binToHex(lcl.hash);
+                }
+
+                // Validator should publish at least two checkpoints
+                return loadDone &&
+                       simulation->haveAllExternalized(3 * checkpoint, 5);
+            },
+            50 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+        auto lcl = simulation->getNode(vNode2NodeID)
+                       ->getLedgerManager()
+                       .getLastClosedLedgerHeader();
+
+        // Shutdown captive core and test various startup paths
+        simulation->removeNode(vNode2NodeID);
+
+        SECTION("minimal DB setup")
+        {
+            // Shutdown captive core
+            SECTION("not found")
+            {
+                // Remove `buckets` dir completely
+                fs::deltree(cfg2.BUCKET_DIR_PATH);
+
+                // Initialize new minimal DB from scratch
+                auto app = setupApp(cfg2, clock, 0, "");
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+            }
+            SECTION("found")
+            {
+                // Found existing minimal DB, reset to genesis
+                auto app = setupApp(cfg2, clock, 0, "");
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+            }
+        }
+        SECTION("rebuild state")
+        {
+            SECTION("from buckets")
+            {
+                selectedLedger = lcl.header.ledgerSeq;
+                selectedHash = binToHex(lcl.hash);
+
+                auto app = setupApp(cfg2, clock, selectedLedger, selectedHash);
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+                REQUIRE(app->getLedgerManager().getLastClosedLedgerNum() ==
+                        selectedLedger);
+            }
+            SECTION("via catchup")
+            {
+                // startAtLedger is behind LCL, reset to genesis and catchup
+                auto app = setupApp(cfg2, clock, selectedLedger, selectedHash);
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+                REQUIRE(app->getLedgerManager().getLastClosedLedgerNum() ==
+                        selectedLedger);
+            }
+        }
+        SECTION("set meta stream")
+        {
```

Comment:
There is a `readLcms` lambda over in src/ledger/tests/LedgerCloseMetaStreamTests.cpp that open a meta stream and extracts its content -- might hoist that out to a common test utility routine and use it here to check that the stream has what you expect.

Reply:
Mmm, we do not write to the stream here though, at this is just to test setup (i.e. spinning up multiple apps on startup shouldn't impact the stream, a bug we had before).

Reply:
Ok. Thought it might be nice to check it produced the right thing, but I guess that's already being checked in the LedgerCloseMetaStreamTests themselves.